src/aten/sycl/UpSample.h

```diff
@@ -0,0 +1,365 @@
+#pragma once
+#include <ATen/core/TensorAccessor.h>
+#include <aten/sycl/Atomics.h>
+
+#include <c10/util/ArrayRef.h>
+#include <c10/util/Optional.h>
+#include <c10/util/OptionalArrayRef.h>
+#include <c10/util/SmallVector.h>
+
+#include <math.h>
+
+namespace at {
+namespace native {
+namespace xpu {
+
+namespace upsample {
+
+TORCH_API c10::SmallVector<int64_t, 3> compute_output_size(
+    c10::IntArrayRef input_size, // Full input tensor size.
+    at::OptionalIntArrayRef output_size,
+    c10::optional<c10::ArrayRef<double>> scale_factors);
+} // namespace upsample
+
+namespace upsample_xpu {
+
+inline c10::optional<double> get_scale_value(
+    c10::optional<c10::ArrayRef<double>> scales,
+    int idx) {
+  if (!scales) {
+    return nullopt;
+  }
+  return scales->at(idx);
+}
+
+} // namespace upsample_xpu
+
+/* TODO: move this to a common place */
+template <typename scalar_t>
+inline scalar_t min(scalar_t a, scalar_t b) {
+  return a < b ? a : b;
+}
+
+template <typename scalar_t>
+inline scalar_t max(scalar_t a, scalar_t b) {
+  return a > b ? a : b;
+}
+
+// NOTE [ Nearest neighbor upsampling kernel implementation ]
+//
+// The nearest neighbor upsampling kernel implementation is symmetrical as
+// expected. We launch kernels with threads mapping to destination tensors where
+// kernels write data to, each thread reads data from the source tensor, this
+// means:
+// 1. In the forward kernel,
+//      src_xxx refers to properties of input tensors;
+//      dst_xxx refers to properties of output tensors;
+//      scale_factor is the ratio of src_size to dst_size;
+// 2. In the backward kernel,
+//      src_xxx refers to properties of grad_output tensors;
+//      dst_xxx refers to properties of grad_input tensors;
+//      scale_factor is the ratio of src_size to dst_size;
+//
+// Because of this, we need to take the reciprocal of the scale defined by
+// upsample layer during forward path. The motivation is to avoid slow
+// division in the kernel code, so we can use faster multiplication instead.
+// This is not necessary during backward path, since the scale_factor is already
+// the reciprocal of corresponding scale_factor used in the forward path due to
+// the swap of source and destination tensor.
+//
+// Similarly, since the mapping from grad_input to grad_output during backward
+// is the reverse of the mapping of output to input, we need to have opposite
+// mapping functions to compute the source index.
+
+// see NOTE [ Nearest neighbor upsampling kernel implementation ]
+template <typename accscalar_t>
+static inline accscalar_t compute_scales_value(
+    const c10::optional<double> scale,
```

Comment:
do we really need `double` for scale? this implies double arithmetic ops... will this be any issue with ATS-M/DG2 (I thought we consider to have only cast operations for double enabled on these platforms by default)?

Reply:
yes, in these two ops, we use this function like this:`compute_scales_value<float>`, the utils align with cuda and ipex

Reply:
@dvrogozh  It's a helper on host, won't impact platform compatibility.

But you are right. We are considering the compatibility for ATS-M/DG2, which is a separate task. Since double is a basic data type of C++/SYCL, regarding SYCL kernels in torch-xpu-ops, we say, they are all general. Double WA is specific for ATSM and DG2.